lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const { ObjectEntries, ObjectKeys, SafeMap, ArrayIsArray } = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
```

Comment:
Can we please add JSDoc to all these methods? Is `raw` a string of JSON, the contents of the import map file?

We might want to have the “load the import map from disk/network” part live outside of this class, and then the constructor expects only the contents of the import map (and whatever other context it needs, like base URL) so that we can potentially provide this class as an API for hook authors to use.

Reply:
> Can we please add JSDoc to all these methods?

Oh yeah, I just don't waste time doing that stuff until the api discussions are done. I don't want to waste time writing jsdocs just to have to change them for each comment on the PR. So I will address this and leave the comment, but I don't plan to push those until the end.

> Is raw a string of JSON, the contents of the import map file?

`raw` is the parsed json. 

> We might want to have the “load the import map from disk/network” part live outside of this class, and then the constructor expects only the contents of the import map

It does currently. It is in run_main. I planned use this in the unit tests to pass in import maps, although the initial ones I wrote used fixtures from the filesystem mostly.